Add cross-links to API docstring from tutorial and user-guide

amanbagrecha opened the pull request:
<!-- Feel free to remove check-list items aren't relevant to your change -->

- [x] Partially Addresses #8008 
- [x] ~Tests added~
- [x] ~User visible changes (including notable bug fixes) are documented in `whats-new.rst`~
- [x] ~New functions/methods are listed in `api.rst`~

Cross links for
- apply_ufunc
- weighted
- indexing
- interp
- map_blocks
- groupby
- coarsen

dcherian commented on xarray/core/computation.py:
```diff
@@ -1141,9 +1141,18 @@ def apply_ufunc(
+    :ref:`dask.automatic-parallelization`
+        Example guide describing :py:func:`apply_ufunc` for 1d array.
+
+    `*Tutorial on apply_ufunc* <https://tutorial.xarray.dev/advanced/apply_ufunc/simple_numpy_apply_ufunc.html>`_
+        Advanced Tutorial on applying numpy function using ``apply_ufunc``
+
+
+    :doc:`/examples/apply_ufunc_vectorize_1d.html`.
```
Let's jsut link to https://tutorial.xarray.dev/advanced/apply_ufunc/apply_ufunc.html , The `examples` link above will eventually be removed.

dcherian replied:
```suggestion
    `*Tutorial on apply_ufunc* <https://tutorial.xarray.dev/advanced/map_blocks/map_blocks.html>`_
        Advanced Tutorial on applying numpy function using ``apply_ufunc``
```

dcherian commented on xarray/core/dataarray.py:
```diff
@@ -5435,6 +5435,9 @@ def map_blocks(
         dask.array.map_blocks, xarray.apply_ufunc, xarray.Dataset.map_blocks
         xarray.DataArray.map_blocks
 
+        `**Tutorial on map_blocks** <https://tutorial.xarray.dev/advanced/map_blocks/simple_map_blocks.html>`_
```
```suggestion
        `**Tutorial on map_blocks** <https://tutorial.xarray.dev/advanced/map_blocks/map_blocks.html>`_
```

dcherian commented on xarray/core/dataset.py:
```diff
@@ -2889,6 +2889,12 @@ def isel(
         DataArray.isel
+
+
+        `**Tutorial on indexing** <https://tutorial.xarray.dev/intermediate/indexing/advanced-indexing.html>`_
+            Intermediate Tutorial on applying advanced indexing on dataset
+
```
```suggestion
        `**Tutorial on indexing** <https://tutorial.xarray.dev/intermediate/indexing/indexing.html>`_
            Tutorial material on indexing with Xarray objects

```

dcherian commented on xarray/core/dataset.py:
```diff
@@ -8633,6 +8639,10 @@ def map_blocks(
         dask.array.map_blocks, xarray.apply_ufunc, xarray.Dataset.map_blocks
         xarray.DataArray.map_blocks
 
+        `**Tutorial on map_blocks** <https://tutorial.xarray.dev/advanced/map_blocks/simple_map_blocks.html>`_
+            Advanced Tutorial on map_blocks with dask
```
```suggestion
        `**Tutorial on map_blocks** <https://tutorial.xarray.dev/advanced/map_blocks/map_blocks.html>`_
            Advanced Tutorial on map_blocks with dask
```

dcherian commented:
Thanks @amanbagrecha . I changed the links to more general pages.

amanbagrecha commented:
Thanks @dcherian for the edits. I reverted the docs you made for `apply_ufunc` to the correct url. Also added additional cross-links to user-guide and tutorials. I have few questions
- Shall I add as many cross-links for the API to user-guide and tutorial in the same PR?
- Currently I am using absolute URL for user-guide links, for eg: https://github.com/pydata/xarray/pull/8311/files#diff-386f35dda14ca12d315f2eb71f0ab35d7fdc73419b5dc5175ea5720f3206903bR6953-R6954, is there a better way to it, do avoid breakage in the future?

keewis reviewed (commented):
I wonder if we should be using the `intersphinx` entries in `https://tutorial.xarray.dev/objects.inv`?

If so, you'd need to add
```python
    "xarray-tutorial": ("https://tutorial.xarray.dev/", None),
```
to `intersphinx_mapping`

amanbagrecha commented on xarray/core/dataarray.py:
```diff
@@ -6788,6 +6801,13 @@ def weighted(self, weights: DataArray) -> DataArrayWeighted:
         --------
         Dataset.weighted
+
+        `Weighted array reductions <https://docs.xarray.dev/en/stable/user-guide/computation.html#weighted-array-reductions>`_
```
many thanks! I made the necessary changes.

amanbagrecha commented:
> I wonder if we should be using the `intersphinx` entries in `https://tutorial.xarray.dev/objects.inv`?
> 
> If so, you'd need to add
> 
> ```python
>     "xarray-tutorial": ("https://tutorial.xarray.dev/", None),
> ```
> 
> to `intersphinx_mapping`

I added xarray-tutorial to intersphinx_mapping. Though I am confused how to use it. I tried `sphobjinv su -u https://tutorial.xarray.dev/ fundamentals` and found the necessary link, though unable to reference it correctly.
I tried the following to no avail. How do I resolve this?
```
:doc:`fundamentals/03.4_weighted`
```

keewis commented:
not sure, you might have to use a prefix:
```rst
:doc:`xarray-tutorial:fundamentals/03.4_weighted`
```
where the prefix is the value you put as the `dict` key in `intersphinx_mapping`

amanbagrecha commented:
> not sure, you might have to use a prefix:
> 
> ```
> :doc:`xarray-tutorial:fundamentals/03.4_weighted`
> ```
> 
> where the prefix is the value you put as the `dict` key in `intersphinx_mapping`

Thanks this worked. Now I'm struggling to find a way to link just the section as in https://tutorial.xarray.dev/intermediate/01-high-level-computation-patterns.html#label-space-windows-or-bins-groupby. Since tutorials are in markdown, I cannot add a directive I think.

keewis commented:
I think this should be possible (not sure though, especially given executablebooks/jupyter-book#1767), but might need changes to the tutorial. @dcherian, any ideas?

dcherian commented:
No ideas but this is a major improvement. So let's just go ahead with a link to the whole page.

keewis reviewed (approved):
we might have to revisit the links once `xarray-tutorial` has section targets, but for now this looks good to me. Thanks a lot, @amanbagrecha

dcherian commented on xarray/core/dataarray.py:
```diff
@@ -1431,6 +1431,12 @@ def isel(
         Dataset.isel
         DataArray.sel
 
+        :doc:`xarray-tutorial:intermediate/indexing/indexing`
```
:( This seems to have messed up some formatting though:
<img width="501" alt="image" src="https://github.com/pydata/xarray/assets/2448579/8f274efb-46c7-489e-8676-ceefb9c851d1">

while on `main`
<img width="476" alt="image" src="https://github.com/pydata/xarray/assets/2448579/05a48127-663f-42e9-b580-0f84f9cfb6e2">

keewis replied:
this should be valid, though. So we might need to fix `napoleon` to allow this.

keewis replied:
it appears we had a similar issue more than two years ago when we first introduced `:doc:` links to `See Also` (see #5176), where we concluded that this is a bug in `napoleon` (nobody got around to reporting / fixing that since then, though).

dcherian replied:
OK well this is an improvement in any case. Merging.

dcherian commented:
Thanks @amanbagrecha this is a great contribution. Welcome to Xarray!
